external/fv3fit/tests/test_tfdataset_ops.py

```diff
@@ -0,0 +1,202 @@
+from fv3fit._shared.config import SliceConfig
+from fv3fit._shared.packer import (
+    pack_tfdataset,
+    unpack_tfdataset,
+    clip_sample,
+    PackingInfo,
+)
+import tensorflow as tf
+from typing import Mapping, Sequence
+import numpy as np
+import pytest
+
+
+def assert_datasets_equal(dataset1, dataset2, rtol=1e-6):
+    sample = next(iter(dataset1))
+    if isinstance(sample, dict):
+        for name in sample:
+            result = tf.data.Dataset.zip((dataset1, dataset2)).map(
+                lambda x1, x2: tf.reduce_any(
+                    tf.abs(x2[name] - x1[name]) / (0.5 * (x2[name] + x1[name])) > rtol
+                )
+            )
+        assert not tf.reduce_any(any(result)), name
+    else:
+        result = tf.data.Dataset.zip((dataset1, dataset2)).map(
+            lambda x1, x2: tf.reduce_any(tf.abs(x2 - x1) / (0.5 * (x2 + x1)) > rtol)
+        )
+        assert not tf.reduce_any(any(result))
+
+
+def get_tfdataset(
+    n_dims: int, n_samples: int, entry_features: Mapping[str, int],
+):
+    if n_dims < 2:
+        raise ValueError(
+            "n_dims must be at least 2 (requires sample, feature dimensions)"
+        )
+    data = {}
+    for name, n_features in entry_features.items():
+        shape = list(range(2, n_dims + 2))
+        shape[-1] = n_features
+        shape[0] = n_samples
+        data[name] = tf.convert_to_tensor(
+            np.random.uniform(size=shape).astype(np.float32)
+        )
+    return get_tfdataset_from_data(data)
+
+
+def get_tfdataset_from_data(data):
+    def gen():
+        yield data
+
+    sample = next(iter(gen()))
+    if isinstance(sample, dict):
+        output_signature = {
+            key: tf.TensorSpec(val.shape, dtype=val.dtype)
+            for key, val in sample.items()
+        }
+    else:
+        output_signature = tf.TensorSpec(sample.shape, dtype=sample.dtype)
+    return tf.data.Dataset.from_generator(
+        gen, output_signature=output_signature
+    ).unbatch()
+
+
+def test_assert_datasets_equal_raises():
+    dataset_1 = get_tfdataset_from_data(tf.convert_to_tensor(np.asarray([0.0])))
+    dataset_2 = get_tfdataset_from_data(tf.convert_to_tensor(np.asarray([1.0])))
+    with pytest.raises(AssertionError):
+        assert_datasets_equal(dataset_1, dataset_2, rtol=0.1)
+
+
+@pytest.mark.parametrize("n_dims", [2, 3, 5])
+@pytest.mark.parametrize(
+    "variable_names",
+    [
+        pytest.param(["a", "b", "c", "d"], id="all"),
+        pytest.param(["b"], id="one"),
+        pytest.param(["a", "c"], id="half"),
+    ],
+)
+def test_pack_tfdataset(n_dims: int, variable_names: Sequence[str]):
```

Comment:
if switching to `_seq_to_tf_dataset` I don't think these transform-specific tests are really necessary.  Mostly testing that they work with `dataset.map` which is controlled by the user and how they order these operations.  If all the data changes happen w/in a composed function we don't have to worry about intermediate tf_dataset states.

Reply:
This is testing things work testing, I think? For example, it's testing that the returned PackingInfo is accurate (most of the code below), and it's testing that the function applied inside `map` is accurate.

With some kind of Pipeline class, you still generally want unit tests to make sure each of the functions composing the Pipeline are correct.

```
def pack_tfdataset(
    data: tf.data.Dataset, variable_names: Sequence[str],
) -> Tuple[tf.data.Dataset, PackingInfo]:
    sample = next(iter(data))
    features = []
    for name in variable_names:
        if len(sample[name].shape) > 0:
            features.append(sample[name].shape[-1])
        else:
            features.append(1)
    packing_info = PackingInfo(names=variable_names, features=features)
    return (
        data.map(
            lambda x: tf.concat(list(x[name] for name in variable_names), axis=-1)
        ),
        packing_info,
    )
```

Reply:
This just seemed like a lot of extra code work to make sure that the tf_dataset plays with the individual transforms. If using another composition method, it would take the tf_dataset-specific code out of the testing mix, which is nice, but I also see why the flexibility of using `map` for individual transforms is also nice.